src/TechCounts.cpp

```diff
@@ -0,0 +1,268 @@
+
+	std::vector<RString> asTechCounts;
+	for( int r=0; r < iMaxValues; r++ )
+	{
+		asTechCounts.push_back(ssprintf("%.3f", (*this)[r]));
+	}
+
+	return join( ",",asTechCounts );
+}
+
+void TechCounts::FromString( RString sTechCounts )
+{
+	std::vector<RString> saValues;
+	split( sTechCounts, ",", saValues, true );
+
+	if( saValues.size() != NUM_TechCountsCategory )
+	{
+		MakeUnknown();
+		return;
+	}
+
+	FOREACH_ENUM( RadarCategory, rc )
+	{
+		(*this)[rc] = StringToFloat(saValues[rc]);
+	}
+
+}
+
+void TechCounts::CalculateTechCountsFromRows(const std::vector<StepParity::Row> &rows, TechCounts &out)
+{
+	// arrays to hold the column for each Foot enum.
+	// A value of -1 means that Foot is not on any column
+	int previousFootPlacement[StepParity::NUM_Foot];
+	int currentFootPlacement[StepParity::NUM_Foot];
+	int previousNoteCount = 0;
+	for (int f = 0; f < StepParity::NUM_Foot; f++)
+	{
+		previousFootPlacement[f] = -1;
+		currentFootPlacement[f] = -1;
+	}
+
+	// arrays to hold the foot placements for the current row and previos row.
+	// They're basically just so we don't have to reference currentRow.notes[c].parity everywhere
+	std::vector<StepParity::Foot> previousColumns(rows[0].columnCount, StepParity::NONE);
+	std::vector<StepParity::Foot> currentColumns(rows[0].columnCount, StepParity::NONE);
+	
+	for (unsigned long i = 1; i < rows.size(); i++)
+	{
+		const StepParity::Row &currentRow = rows[i];
+		const StepParity::Row &previousRow = rows[i - 1];
+			
+		int noteCount = 0;
+		
+		// copy the foot placement for the current row into currentColumns,
+		// and count up how many notes there are in this row
+		for (int c = 0; c < currentRow.columnCount; c++)
+		{
+			StepParity::Foot currFoot = currentRow.notes[c].parity;
+			TapNoteType currType = currentRow.notes[c].type;
+
+			// If this isn't either a tap or the beginning of a hold, skip it
+			if(currType != TapNoteType_Tap && currType != TapNoteType_HoldHead)
+			{
+				continue;
+			}
+
+			currentFootPlacement[currFoot] = c;
+			currentColumns[c] = currFoot;
+			noteCount += 1;
+		}
+
+		/*
+		Jacks are same arrow same foot
+		Doublestep is same foot on successive arrows
+		Brackets are jumps with one foot
+
+		Footswitch is different foot on the up or down arrow
+		Sideswitch is footswitch on left or right arrow
+		Crossovers are left foot on right arrow or vice versa
+		*/
+
+		// check for jacks and doublesteps
+		if(noteCount == 1 && previousNoteCount == 1)
+		{
+			for (StepParity::Foot foot: StepParity::FEET)
+			{
+				if(currentFootPlacement[foot] == -1 || previousFootPlacement[foot] == -1)
+				{
+					continue;
+				}
+				
+				if(previousFootPlacement[foot] == currentFootPlacement[foot])
+				{
+					if(currentRow.second - previousRow.second < JACK_CUTOFF)
+					{
+						out[TechCountsCategory_Jacks] += 1;
+					}
+				}
+				else
+				{
+					out[TechCountsCategory_Doublesteps] += 1;
+				}
+			}
+		}
+
+		// check for brackets
+		if(noteCount >= 2)
```

Comment:
So what I think is maybe happening is this logic is only executing if `noteCount == 1 && previousNoteCount == 1`. If we have consecutive jumps (`noteCount >=2`) we are not processing them with the correct logic?